### DISCUSSION TITLE
close the application

### DISCUSSION THREAD

**Discussion by hulkdx** [arkivanov/Decompose#738] [Q&A]
Is it possible to close the application with decompose?
I noticed that navigation.pop() does not close the application if backstack is empty how to do that?

**Comment by arkivanov** [comment]
The general idea is to leverage DI. One of the ways is to pass a callback from the root Activity (or equivalent thing on other platforms) down to a component, and call `finish` or something when the callback is called. Another way is to define an interface that would close the app, and inject it via DI.

**Reply by hulkdx** [reply]
@arkivanov Thanks, is it possible that you provide some samples for both options?
I ended up doing something like this:
https://github.com/hulkdx/findprofessional-frontend-mobile/blob/main/libs/navigation-decompose/src/commonMain/kotlin/com/hulkdx/findprofessional/libs/navigation/decompose/DecomposeNavigator.kt#L39-L45

Which uses exitProcess(0) on android
https://github.com/hulkdx/findprofessional-frontend-mobile/blob/main/core/src/androidDebug/kotlin/com/hulkdx/findprofessional/core/config/PlatformSpecific.android.kt#L23-L25

I find it a bit difficult to finish the activity, if you could provide some samples would be really helpful.

**Reply by arkivanov** [reply]
Sure! You can pass a callback to your root component, and any other components down the tree that need to trigger the action. On each level you can decide wether to pop the stack or to bubble the event up, until it finally reaches the root.

```kotlin
class DefaultRootComponent(
    componentContext: ComponentContext,
    private val onFinished: () -> Unit,
) : RootComponent, ComponentContext by componentContext
```

Then in your `MainActivity` you can just call `finish`.

```kotlin
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val root =
            DefaultRootComponent(
                componentContext = defaultComponentContext(),
                onFinish = ::finish,
            )

        setContent {
            ...
        }
    }
}
```

Also, since you seem to use KMP, there looks like [no proper way](https://stackoverflow.com/questions/355168/proper-way-to-exit-iphone-application) of closing the app programmatically.
